README
NAME
    MooseX::Enumeration - a native attribute trait for enums

SYNOPSIS
    Given this class:

       package MyApp::Result {
          use Moose;
          use Types::Standard qw(Enum);
          has status => (
             is        => "rw",
             isa       => Enum[qw/ pass fail /],
          );
       }

    It's quite common to do this kind of thing:

       if ( $result->status eq "pass" ) { ... }

    But if you're throwing strings around, it can be quite easy to mistype
    them:

       if ( $result->status eq "apss" ) { ... }

    And the comparison silently fails. Instead, let's define the class like
    this:

       package MyApp::Result {
          use Moose;
          use Types::Standard qw(Enum);
          has status => (
             traits    => ["Enumeration"],
             is        => "rw",
             isa       => Enum[qw/ pass fail /],
             handles   => [qw/ is_pass is_fail /],
          );
       }

    So you can use the class like this:

       if ( $result->is_pass ) { ... }

    Yay!

DESCRIPTION
    This attribute trait makes it easier to work with enumerated types in
    Moose. (Hopefully a Moo implementation will follow soon.)

    It will only work on attributes which have an enum type constraint. This
    may be a Type::Tiny::Enum or may be a type constraint defined using
    Moose's built-in enum types.

  Type Constraint Shortcut
    This trait gives you a shortcut for specifying an enum type constraint:

       has status => (
          traits    => ["Enumeration"],
          is        => "rw",
          enum      => [qw/ pass fail /],   # instead of isa
       );

  Delegation
    The trait also allows you to delegate "is" to the attribute value.

       # the most longhanded form...
       #
       has status => (
          traits    => ["Enumeration"],
          is        => "rw",
          enum      => [qw/ pass fail /],
          handles   => {
             is_pass  => ["is", "pass"],
             is_fail  => ["is", "fail"],
          }
       );

    Note that above, we might have called the delegated method "did_pass"
    instead of "is_pass". You can call it what you like.

    To save typing, we offer some shorthands for common patterns.

       has status => (
          traits    => ["Enumeration"],
          is        => "rw",
          enum      => [qw/ pass fail /],
          handles   => {
             is_pass  => "is_pass",
             is_fail  => "is_fail",
          }
       );

    In the hashref values, we implicitly split on the first underscore, so
    "is_pass" is equivalent to `["is", "pass"]`.

    This is still repetitive, so how about...

       has status => (
          traits    => ["Enumeration"],
          is        => "rw",
          enum      => [qw/ pass fail /],
          handles   => [ "is_pass", "is_fail" ],
       );

    If an arrayref of delegates is given, it mapped like this:

       my %delegate_hash = map { $_ => $_ } @delegate_array;

    We can still go one better...

       has status => (
          traits    => ["Enumeration"],
          is        => "rw",
          enum      => [qw/ pass fail /],
          handles   => 1,
       );

    This will create a delegated method for each value in the enumeration.

PERFORMANCE
    As of version 0.003, `$obj->is_pass` actually benchmarks *faster* than
    `$obj->status eq "pass"`. The latter comparison can be accelerated using
    MooseX::XSAccessor but this module can not (yet) provide an XS version for
    `is_pass`. :-(

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-Enumeration>.

SEE ALSO
    Moose::Meta::TypeConstraint::Enum, Type::Tiny::Enum,
    Moose::Meta::Attribute::Native.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
MooseX-Enumeration
==================

Created:      2014-05-16
Home page:    <https://metacpan.org/release/MooseX-Enumeration>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-Enumeration>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.004	2014-05-19

 [ Bug Fixes ]
 - Avoid role conflict on Moose older than 2.0800.

0.003	2014-05-18

 [ Packaging ]
 - Declare dependency on Module::Runtime.

 [ Other ]
 - Improved efficiency of delegated methods; they now run at twice the
   speed!
 - No longer require Type::Tiny::Enum; fall back to
   Moose::Meta::TypeConstraint::Enum.

0.002	2014-05-16

 [ Test Suite ]
 - Add a test case for non-Type::Tiny type constraint.

 [ Packaging ]
 - Declare test dependency on Types::Standard.

0.001	2014-05-16	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "a native attribute trait for enums",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.015, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "MooseX-Enumeration",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       recommends => { "Type::Tiny::Enum" => "0.030" },
                                       requires   => { "Module::Runtime" => 0, "Moose" => 1.15, "perl" => 5.008001 },
                                     },
                        test      => { requires => { "Test::More" => 0.92 } },
                      },
  "provides"       => {
                        "Moose::Meta::Attribute::Custom::Trait::Enum" => {
                          file => "lib/Moose/Meta/Attribute/Custom/Trait/Enum.pm",
                          version => 0.004,
                        },
                        "Moose::Meta::Attribute::Custom::Trait::Enumeration" => {
                          file => "lib/Moose/Meta/Attribute/Custom/Trait/Enumeration.pm",
                          version => 0.004,
                        },
                        "MooseX::Enumeration" => { file => "lib/MooseX/Enumeration.pm", version => 0.004 },
                        "MooseX::Enumeration::Meta::Attribute::Native::Trait::Enumeration" => {
                          file => "lib/MooseX/Enumeration/Meta/Attribute/Native/Trait/Enumeration.pm",
                          version => 0.004,
                        },
                        "MooseX::Enumeration::Meta::Method::Accessor::Native::Enumeration::is" => {
                          file => "lib/MooseX/Enumeration/Meta/Method/Accessor/Native/Enumeration/is.pm",
                          version => 0.004,
                        },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=MooseX-Enumeration",
                                        },
                        homepage     => "https://metacpan.org/release/MooseX-Enumeration",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-moosex-enumeration.git",
                                          web  => "https://github.com/tobyink/p5-moosex-enumeration",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/MooseX-Enumeration/project",
                      },
  "version"        => 0.004,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

